typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -1077,92 +1077,141 @@
+  (-> (listof symbol?) Rep? Rep?)
+  (define n-1 (sub1 (length names-to-abstract)))
+  (define (abstract-name name lvl default dotted?)
+    (match (index-of names-to-abstract name eq?)
+      [#f default]
+      ;; adjust index properly before using (see comments above
+      ;; and note we are under 'lvl' additional binders)
+      [idx (let ([idx (+ lvl (- n-1 idx))])
+             (cond [dotted? idx]
+                   [else (make-B idx)]))]))
+  (let rec/lvl ([cur initial] [lvl 0])
+    (define (rec rep) (rec/lvl rep lvl))
+    (match cur
+      ;; Free type variables
+      [(F: name) (abstract-name name lvl cur #f)]
+      ;; forms w/ (potential) dotted type variables
+      [(arr: dom rng rest drest kws)
+       (make-arr (map rec dom)
+                 (rec rng)
+                 (and rest (rec rest))
+                 (match drest
+                   [(cons d-ty d)
+                    (cons (rec d-ty) (abstract-name d lvl d #t))]
+                   [_ #f])
+                 (map rec kws))]
+      [(ValuesDots: rs dty d)
+       (make-ValuesDots (map rec rs)
+                        (rec dty)
+                        (abstract-name d lvl d #t))]
+      [(ListDots: dty d)
+       (make-ListDots (rec dty)
+                      (abstract-name d lvl d #t))]
+      ;; forms which introduce bindings (increment lvls appropriately)
+      [(Mu: body) (make-Mu (rec/lvl body (add1 lvl)))]
+      [(PolyRow: constraints body)
+       (make-PolyRow constraints (rec/lvl body (add1 1 lvl)))]
+      [(PolyDots: n body)
+       (make-PolyDots n (rec/lvl body (+ n lvl)))]
+      [(Poly: n body)
+       (make-Poly n (rec/lvl body (+ n lvl)))]
+      [_ (Rep-fmap cur rec)])))
+
+
+;; instantiate
+;;
+;; instantiates De Bruijn index 0 with
+;; 'type' in 'initial'
+(define/cond-contract (instantiate type initial)
+  (-> Type? Rep? Rep?)
+  (instantiate-many (list type) initial))
+
+;; instantiate-many
+;;
+;; instantiates De Bruijn indices 0 through
+;; (sub1 (length images)) with the types from images.
+;; Specifically, if n = (length images), then
+;; index 0 gets mapped to images[n-1]
+;; index 1 gets mapped to images[n-2]
+;; ...
+;; index n-1 gets mapped to images[0]
+(define/cond-contract (instantiate-many images initial)
+  (-> (listof Type?) Rep? Rep?)
+  (define n-1 (sub1 (length images)))
+  (define (instantiate-idx idx lvl default dotted?)
+    ;; adjust for being under 'lvl' binders and for
+    ;; index 0 gets mapped to images[n-1], etc
+    (let ([idx (- n-1 (- idx lvl))])
+      (match (list-ref/default images idx #f)
+        [#f default]
+        [image (cond [dotted? (F-n image)]
+                     [else image])])))
+  (let rec/lvl ([cur initial] [lvl 0])
+    (define (rec rep) (rec/lvl rep lvl))
+    (match cur
+      ;; Bound De Bruijn indices
+      [(B: idx) (instantiate-idx idx lvl cur #f)]
+      ;; forms w/ (potential) dotted type indices
+      [(arr: dom rng rest (cons d-ty (? exact-integer? d)) kws)
```

Comment:
yes that behavior should be correct, `arr` structs know how to generically map over themselves correctly